### DISCUSSION TITLE
View domains layout

### DISCUSSION THREAD

**Discussion by ajagnanan** [DigiLive/mushroom-strategy#59] [General]
I've been working on getting views setup for each tablet on the wall. The layout and ordering of Entities on views keep changing, so sometimes ordering is Entity A then B, other times B then A. The Domains can sometimes be out of order too (e.g. Sensors before Switches then Switches before Sensors) I haven't been able to pinpoint what's causing this. I did figure out that Mushroom layout (vertical and horizontal) contain the Entities. Maybe this is an issue with Mushroom Template itself.

**Comment by DigiLive** [comment]
Can you clarify what you mean?
I don't really understand what you're asking.

**Reply by ajagnanan** [reply]
@DigiLive updated the text

**Comment by DigiLive** [comment]
I've had a look into this and it seems we're ordering the entities on an inappropriate value.

The following should apply:

- Entities are sorted by `original_name` instead of `friendly_name` (which is a property of an entity's state object).
  (`original_name` is often `null`, which could explain the "random" order.)

- Areas are sorted by `name`.

- Views (at the navigation bar) are sorted by `title`.
  (Overruled by the `order` property at `optionDefaults.js`)

- Domains are sorted by `title`.
  (The default domain (Miscellaneous) always listed as last.

All of the above sorting-orders are overruled if a order property is defined. In that case the value of this property is used to sort.

It's HA which is placing our (horizontal/vertical stacked) cards into columns.
I don't know the logic behind it, but you should see the correct ordering (according to above's explanation) when making the client's width smaller until it leaves you with 1 column.

**Reply by ajagnanan** [reply]
Yea, I'm seeing the same. HA is laying out the vertical/horizontal stacked cards and columns. I think its float left controlled, so if a previous column isn't long enough, some domains flow to that column and disrupts the ordering. I have been using the `custom:stack-in-card` to get some placement control. Unfortunately I'm at customizing 75% of a view via the manual method with `extra_cards`. There's 2 things I can see so far that will help:

- Sort the entities by Id? friendly_name might work too, but it can change ordering.
- Can we add the ability to place extra_cards before and after what's generated?

**Reply by DigiLive** [reply]
I think we can never guarantee a correct order because of the way HA places the generated stacks and re-places them on other widths. 
Maybe we should make use of a [Grid Card](https://www.home-assistant.io/dashboards/grid/) or something.
FYI: It's not float-left, but flexbox-row

> Sort the entities by Id? friendly_name might work too, but it can change ordering.

It should be `friendly_name` instead of `original_name` which is used now.
If you have concerns about sorting entities with the same name, you could concatenate the id to the name (just) for ordering.

> Can we add the ability to place extra_cards before and after what's generated?

I'm not sure what you mean.

**Reply by ajagnanan** [reply]
>Can we add the ability to place extra_cards before and after what's generated?

Currently cards are placed in a List like this: 
`extra_cards` then `generated cards`

Can we add a new property `extra_cards_after` that makes it this:
`extra_cards` then `generated cards` then `extra_cards_after`

This way we can place custom cards before and after the generated cards.

**Reply by DigiLive** [reply]
It think doing an `extra_cards_before` and `extra_cards_after` shouldn't be a problem.
You can create an [issue ](https://github.com/AalianKhan/mushroom-strategy/issues/new?assignees=&labels=enhancement&projects=&template=feature_request.md&title=Add%20option%20extra_cards_before) for it.

**Reply by ajagnanan** [reply]
Done https://github.com/AalianKhan/mushroom-strategy/issues/61, thanks guys

**Comment by ajagnanan** [comment]
Btw, congrats guys! The project was mentioned in https://youtu.be/LDq5A-bYxT8?t=389
